File: java/Algorithms/src/Collections/MaxPriorityQueue.java
package Collections;

import java.util.Comparator;
import java.util.Iterator;
import java.util.NoSuchElementException;

/**
 * Created by jianming.xiao on 11/9/14.
 */
public class MaxPriorityQueue<Key> extends PriorityQueue<Key> {
    /**
     * Initializes an empty priority queue.
     */
    public MaxPriorityQueue() {
        super();
    }

    /**
     * Initializes an empty priority queue with the given initial capacity.
     *
     * @param initCapacity the initial capacity of the priority queue
     */
    public MaxPriorityQueue(int initCapacity) {
        super(initCapacity);
    }

    /**
     * Initializes an empty priority queue using the given comparator.
     *
     * @param comparator the order in which to compare the keys
     */
    public MaxPriorityQueue(Comparator<Key> comparator) {
        super(comparator);
    }

    /**
     * Initializes an empty priority queue with the given initial capacity,
     * using the given comparator.
     *
     * @param initCapacity the initial capacity of the priority queue
     * @param comparator   the order in which to compare the keys
     */
    public MaxPriorityQueue(int initCapacity, Comparator<Key> comparator) {
        super(initCapacity, comparator);
    }

    /**
     * Initializes a priority queue from the array of keys.
     * Takes time proportional to the number of keys, using sink-based heap construction.
     *
     * @param keys the array of keys
     */
    public MaxPriorityQueue(Key[] keys) {
        super(keys);
    }

    /**
     * Returns a largest key on the priority queue.
     *
     * @return a largest key on the priority queue
     * @throws java.util.NoSuchElementException if the priority queue is empty
     */
    public Key max() {
        return first();
    }

    /**
     * Removes and returns a largest key on the priority queue.
     *
     * @return a largest key on the priority queue
     * @throws java.util.NoSuchElementException if priority queue is empty.
     */
    public Key delMax() {
        return delete();
    }

    /**
     * Returns an iterator that iterates over the keys on the priority queue
     * in descending order.
     * The iterator doesn't implement <tt>remove()</tt> since it's optional.
     *
     * @return an iterator that iterates over the keys in descending order
     */
    @Override
    public Iterator<Key> iterator() {
        return new HeapIterator();
    }


    @Override
    protected void sink(int k) {
        while (2 * k <= N) {
            int j = 2 * k;
            if (j < N && less(j, j + 1)) j++;
            if (!less(k, j)) break;
            exch(k, j);
            k = j;
        }
    }

    @Override
    protected void swim(int k) {
        while (k > 1 && less(k / 2, k)) {
            exch(k / 2, k);
            k = k / 2;
        }
    }


    private boolean less(int i, int j) {
        if (comparator == null)
            return ((Comparable<Key>) pq[i]).compareTo(pq[j]) < 0;
        else
            return comparator.compare(pq[i], pq[j]) < 0;
    }

    private class HeapIterator implements Iterator<Key> {
        private MaxPriorityQueue<Key> copy;

        private HeapIterator() {
            if (comparator == null)
                copy = new MaxPriorityQueue<Key>(size());
            else
                copy = new MaxPriorityQueue<Key>(size(), comparator);
            for (int i = 1; i <= N; i++)
                copy.insert(pq[i]);
        }

        @Override
        public boolean hasNext() {
            return !copy.isEmpty();
        }

        @Override
        public Key next() {
            if (!hasNext()) throw new NoSuchElementException();
            return copy.delMax();
        }

        @Override
        public void remove() {
            throw new UnsupportedOperationException();
        }
    }
}

File: c#/BigEgg.Algorithm/BigEgg.Algorithm/Collections/MaxPriorityQueue.cs
using System;
using System.Collections.Generic;

namespace BigEgg.Algorithm.Collections
{
    public class MaxPriorityQueue<Key> : PriorityQueue<Key>
    {
        /// <summary>
        /// Initializes an empty priority queue.
        /// </summary>
        public MaxPriorityQueue()
            : base()
        { }

        /// <summary>
        /// Initializes an empty priority queue with the given initial capacity.
        /// </summary>
        /// <param name="initCapacity">The initial capacity of the priority queue</param>
        public MaxPriorityQueue(int initCapacity)
            : base(initCapacity)
        { }

        /// <summary>
        /// Initializes an empty priority queue using the given comparator.
        /// </summary>
        /// <param name="comparer">The order in which to compare the keys.</param>
        public MaxPriorityQueue(IComparer<Key> comparer)
            : base(comparer)
        { }

        /// <summary>
        /// Initializes an empty priority queue with the given initial capacity,
        /// using the given comparator.
        /// </summary>
        /// <param name="initCapacity">The initial capacity of the priority queue.</param>
        /// <param name="comparer">The order in which to compare the keys.</param>
        public MaxPriorityQueue(int initCapacity, IComparer<Key> comparer)
            : base(initCapacity, comparer)
        { }

        /// <summary>
        /// Initializes a priority queue from the array of keys.
        /// Takes time proportional to the number of keys, using sink-based heap construction.
        /// </summary>
        /// <param name="keys">The array of keys</param>
        public MaxPriorityQueue(Key[] keys)
            : this(keys, null)
        { }

        /// <summary>
        /// Initializes a priority queue from the array of keys.
        /// Takes time proportional to the number of keys, using sink-based heap construction.
        /// </summary>
        /// <param name="keys">The array of keys</param>
        /// <param name="comparer">The order in which to compare the keys.</param>
        public MaxPriorityQueue(Key[] keys, IComparer<Key> comparer)
            : base(keys, comparer)
        { }


        /// <summary>
        /// Returns the largest key on the priority queue.
        /// </summary>
        /// <returns>The largest key on the priority queue</returns>
        /// <exception cref="System.ArgumentOutOfRangeException">Queue is empty.</exception>
        public Key Max()
        {
            return First();
        }

        /// <summary>
        /// Removes and returns the largest key on the priority queue.
        /// </summary>
        /// <returns>The largest key on the priority queue.</returns>
        /// <exception cref="System.ArgumentOutOfRangeException">Queue is empty.</exception>
        public Key DelMax()
        {
            return Delete();
        }

        /// <summary>
        /// Returns an enumerator that iterates over the keys on the priority queue
        /// in descending order.
        /// </summary>
        /// <returns>
        /// A <see cref="T:System.Collections.Generic.IEnumerator`1" /> that can be used to iterate over the keys.
        /// </returns>
        public override IEnumerator<Key> GetEnumerator()
        {
            return new HeapEnumerator(this);
        }


        protected override void Sink(int k)
        {
            while (2 * k <= N)
            {
                int j = 2 * k;
                if (j < N && Less(j, j + 1)) j++;
                if (!Less(k, j)) break;
                Exch(k, j);
                k = j;
            }
        }

        protected override void Swim(int k)
        {
            while (k > 1 && Less(k / 2, k))
            {
                Exch(k / 2, k);
                k = k / 2;
            }
        }


        private bool Less(int i, int j)
        {
            if (comparer == null)
                return ((IComparable)pq[i]).CompareTo(pq[j]) < 0;
            else
                return comparer.Compare(pq[i], pq[j]) < 0;
        }

        private class HeapEnumerator : IEnumerator<Key>
        {
            MaxPriorityQueue<Key> copy;

            public HeapEnumerator(MaxPriorityQueue<Key> pq)
            {
                if (pq.comparer == null)
                    copy = new MaxPriorityQueue<Key>(pq.Size());
                else
                    copy = new MaxPriorityQueue<Key>(pq.Size(), pq.comparer);
                for (int i = 1; i <= pq.N; i++)
                {
                    copy.Insert(pq.pq[i]);
                }
            }


            public Key Current
            {
                get { return copy.DelMax(); }
            }

            public void Dispose()
            {
                copy = null;
            }

            object System.Collections.IEnumerator.Current
            {
                get { return Current; }
            }

            public bool MoveNext()
            {
                return !copy.IsEmpty();
            }

            public void Reset()
            {
                throw new NotImplementedException();
            }
        }
    }
}
